src/workshops/gradients/gradientEditor.ts

```diff
@@ -2,54 +2,128 @@ import * as ko from "knockout";
 
-        this.colorStops.push({ color: "#AAAAAA", length: colorStopLength });
-        this.applyChanges();
+        gradient.displayName.subscribe(() => {
+            this.applyChanges();
+            this.updateOnSelect();
+        });
+        gradient.direction.subscribe(this.applyChanges);
+        gradient.colorStops().forEach((colorStop) => {
+            colorStop.color.subscribe(this.applyChanges);
+            colorStop.length.subscribe(this.applyChanges);
+        });
+    }
+
+    public initializePointer(element: HTMLElement, colorStop: ColorStopViewModel): boolean {
+        this.sliderWidth = element.parentElement.getBoundingClientRect().width;
+        const length = colorStop.length();
+        const position = this.sliderWidth * 1.0 / 100 * length - 4;
+        element.style.left= position + "px";
+        element.style.backgroundColor = colorStop.color();
+        return true;
+    } 
+
+    public async addColor(): Promise<void> {
+        const newColor = new ColorStopViewModel(<LinearGradientColorStopContract> {
+            color: "#000000",
+            length: 0
+        })
+        newColor.color.subscribe(this.applyChanges);
+        newColor.length.subscribe(this.applyChanges);
+        this.gradientViewModel().colorStops.push(newColor);
+        this.updateOnSelect();
     }
 
     private async applyChanges(): Promise<void> {
-        const abc = getLinearGradientString({ key: "", displayName: "", direction: `${this.direction()}deg`, colorStops: this.colorStops() });
+        const gradient = this.selectedGradient();
+        const colorStops: LinearGradientColorStopContract[] = [];
 
-        this.gradientPreview({
-            gradientPreview: {
-                background: { image: abc }
-            }
+        gradient.displayName = this.gradientViewModel().displayName();
+        gradient.direction = parseFloat(this.gradientViewModel().direction()) + "deg";
+        this.gradientViewModel().colorStops().forEach((colorStop) => {
+            colorStops.push(<LinearGradientColorStopContract>{
+                color: colorStop.color(),
+                length: colorStop.length()
+            });
         });
+
+        gradient.colorStops = colorStops;
+        this.updateBackground();
+    }
+
+    private async updateBackground(): Promise<void> {
+        const style = new Style("gradient-preview");
+        style.addRules([new StyleRule("backgroundImage", getLinearGradientString(this.selectedGradient()))]);
+
+        const styleSheet = new StyleSheet();
+        styleSheet.styles.push(style);
+
+        this.gradientPreview(styleSheet);
+    }
+
+    public changeColor(obIndex: ko.Observable<number>, colorValue: string): void {
+        const index = obIndex();
+        this.gradientViewModel().colorStops()[index].color(colorValue);
+        this.updateOnSelect();
+    }
+
+
+    public onMouseMove(element: HTMLElement, position: number, index: number): void {
```

Comment:
Tony, is this still needed here? As I mentioned, we should not access the DOM from view models, only bindings should do that :)

Reply:
Oh this is for the logic, so I attached the binding in html for slider into this slider: {onmousemove: $component.onMouseMove}. I made the bindingHandler generic so I need to pass the function in to change the color.